sources/web/datalab/jupyter.ts

```diff
@@ -64,14 +66,19 @@ var templates: common.Map<string> = {
  */
 var appSettings: common.Settings;
 
-function pipeOutput(stream: NodeJS.ReadableStream, port: number, error: boolean) {
+function pipeOutput(server: JupyterServer, stream: NodeJS.ReadableStream, port: number, error: boolean) {
   stream.setEncoding('utf8');
+  var listeningMessage = `The Jupyter Notebook is running at: http://127.0.0.1:${port}`;
+
   stream.on('data', (data: string) => {
     // Jupyter generates a polling kernel message once every 3 seconds
     // per kernel! This adds too much noise into the log, so avoid
     // logging it.
 
     if (data.indexOf('Polling kernel') < 0) {
+      if (data.indexOf(listeningMessage) >= 0) {
```

Comment:
Now we have two checks --- make sure stdout prints the expected message, and make sure the tcp port is used. Having both is perhaps too much and making the code too complicated.

I still think wait_for_port_free --> start_jyputer_process -> wait_for_port_used is better than relying on stdout message (is it possible the message appears before we start listening to it?). But will leave it to you.